README
# Vramsteg ('progress' in Svenska, almost)

Vramsteg is an open source, command line utility that provides shell scripts
with a full-featured progress indicator. The progress bar can display elapsed
time, remaining time estimate, percentage completed, and labels. The progress
bar can have user-specified colors, and be rendered in multiple styles.

See the man page (enter 'man vramsteg') for full details on how to use vramsteg
or look in the srv/examples directory for several shell scripts that illustrate
usage.

---

# -- MIRROR README --

The goal of this repository is to host vramsteg on Github, so that it can be
installed as Zshell plugin, with following commands added to ~/.zshrc:

zi ice as'command' pick'src/vramsteg' atclone'cmake .' atpull'%atclone' make
zi light z-shell/vramsteg-zsh


NEWS
New Features in vramsteg 1.0.1

- Bar labels
- Percentage
- Elapsed time
- Remaining time
- Width setting
- Three styles - default, mono and text

New commands in vramsteg 1.0.1

-

New configuration options in vramsteg 1.0.1

-

Newly deprecated features in vramsteg 1.0.1

- ***

Please refer to the ChangeLog file for full details.

---

Vramsteg has been built and tested on the following configurations:

- OS X 10.8 Mountain Lion and 10.7 Lion
- Debian 7.0 Wheezy (stable) and 6.0 Squeeze (oldstable) plus Sid (testing)

---

While Vramsteg has undergone testing, bugs are sure to remain. If you
encounter a bug, please send a message to:

support@yootabory.com

Thank you.


src/Progress.h
////////////////////////////////////////////////////////////////////////////////
// Vramsteg - Utility for displaying progress bars in shell scripts.
//
// Copyright 2010 - 2015, Paul Beckingham, Federico Hernandez.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
// http://www.opensource.org/licenses/mit-license.php
//
////////////////////////////////////////////////////////////////////////////////

#ifndef INCLUDED_PROGRESS
#define INCLUDED_PROGRESS

#include <string>
#include <time.h>

class Progress
{
public:
  Progress ();
  Progress (const std::string&, int, long, long, bool = true, bool = true);
  ~Progress ();

  void setStyle (const std::string&);
  void setLabel (const std::string&);
  void setWidth (int);
  void setMin (long);
  void setMax (long);
  void showPercentage (bool);
  void removeAfter (bool);
  void setStart (time_t);
  void showEstimate (bool);
  void showElapsed (bool);

  void update (long);
  void done ();

private:
  std::string formatTime (time_t);

  void renderStyleDefault ();
  void renderStyleMono ();
  void renderStyleText ();

private:
  std::string style;
  std::string label;
  int width;
  long minimum;
  long maximum;
  long current;
  bool percentage;
  bool remove;
  time_t start;
  bool estimate;
  bool elapsed;
};

#endif

////////////////////////////////////////////////////////////////////////////////


src/Progress.cpp
////////////////////////////////////////////////////////////////////////////////
// Vramsteg - Utility for displaying progress bars in shell scripts.
//
// Copyright 2010 - 2015, Paul Beckingham, Federico Hernandez.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
// http://www.opensource.org/licenses/mit-license.php
//
////////////////////////////////////////////////////////////////////////////////

#include <iostream>
#include <sstream>
#include <iomanip>
#include <stdio.h>
#include <unistd.h>
#include <Progress.h>

////////////////////////////////////////////////////////////////////////////////
Progress::Progress ()
: style ("")
, label ("")
, width (0)
, minimum (0)
, maximum (0)
, current (-1)
, percentage (true)
, remove (true)
, start (0)
, estimate (false)
, elapsed (false)
{
}

////////////////////////////////////////////////////////////////////////////////
Progress::Progress (const std::string& l, int w, long n, long x, bool p /* = true */, bool r /* = true */)
: style ("")
, label (l)
, width (w)
, minimum (n)
, maximum (x)
, percentage (p)
, remove (r)
, start (0)
, estimate (false)
, elapsed (false)
{
  current = -1;
}

////////////////////////////////////////////////////////////////////////////////
Progress::~Progress ()
{
}

////////////////////////////////////////////////////////////////////////////////
void Progress::setStyle (const std::string& value)
{
  style = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::setLabel (const std::string& value)
{
  label = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::setWidth (int value)
{
  width = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::setMin (long value)
{
  minimum = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::setMax (long value)
{
  maximum = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::showPercentage (bool value)
{
  percentage = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::removeAfter (bool value)
{
  remove = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::setStart (time_t value)
{
  start = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::showEstimate (bool value)
{
  estimate = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::showElapsed (bool value)
{
  elapsed = value;
}

////////////////////////////////////////////////////////////////////////////////
void Progress::update (long value)
{
  if (isatty (fileno (stdout)) && current != value)
  {
    // Box the range.
    if (value < minimum) value = minimum;
    if (value > maximum) value = maximum;

    // Current value.
    current = value;

    // Capable of supporting multiple styles.
         if (style == "")     renderStyleDefault ();
    else if (style == "mono") renderStyleMono ();
    else if (style == "text") renderStyleText ();
    else
      throw std::string ("Style '") + style + "' not supported.";
  }
}

////////////////////////////////////////////////////////////////////////////////
void Progress::done ()
{
  if (isatty (fileno (stdout)))
  {
    if (remove)
      std::cout << "\r"
                << std::setfill (' ')
                << std::setw (width)
                << ' ';

    std::cout << std::endl;
  }
}

////////////////////////////////////////////////////////////////////////////////
std::string Progress::formatTime (time_t t)
{
  char buffer [128];

  int days    =  t          / 86400;
  int hours   = (t % 86400) / 3600;
  int minutes = (t %  3600) / 60;
  int seconds =  t % 60;

  if (days)
    snprintf (buffer, 128, "%dd %d:%02d:%02d", days, hours, minutes, seconds);
  else if (hours)
    snprintf (buffer, 128,     "%d:%02d:%02d",       hours, minutes, seconds);
  else
    snprintf (buffer, 128,        "%02d:%02d",              minutes, seconds);

  return std::string (buffer);
}

////////////////////////////////////////////////////////////////////////////////
// Default style looks like this:
//
// label GGGGGGGGRRRRRRRRRRRRRRRRR  34% 0:12 0:35
//
// ^^^^^                                            Label string
//       ^^^^^^^^                                   Completed bar (in green)
//               ^^^^^^^^^^^^^^^^^                  Incomplete bar (in red)
//                                 ^^^^             Percentage complete
//                                      ^^^^        Elapsed time
//                                           ^^^^   Remaining estimate
void Progress::renderStyleDefault ()
{
  // Fraction completed.
  double fraction = (1.0 * (current - minimum)) / (maximum - minimum);

  // Elapsed time.
  time_t now = time (NULL);
  std::string elapsed_time;
  if (elapsed && start != 0)
    elapsed_time = formatTime (now - start);

  // Estimated remaining time.
  std::string estimate_time;
  if (estimate && start != 0)
  {
    if (fraction >= 1e-6)
      estimate_time = formatTime ((time_t) (int) (((now - start) * (1.0 - fraction)) / fraction));
    else
      estimate_time = formatTime (0);
  }

  // Calculate bar width.
  int bar = width
          - (label.length () ? label.length () + 1         : 0)
          - (percentage      ? 5                           : 0)
          - (elapsed         ? elapsed_time.length () + 1  : 0)
          - (estimate        ? estimate_time.length () + 1 : 0);

  if (bar < 1)
    throw std::string ("The specified width is insufficient.");

  int visible = (int) (fraction * bar);

  // Render.
  if (label.length ())
    std::cout << label
              << ' ';

  if (visible > 0)
    std::cout << "\033[42m" // Green
              << std::setfill (' ')
              << std::setw (visible)
              << ' ';

  if (bar - visible > 0)
    std::cout << "\033[41m" // Red
              << std::setfill (' ')
              << std::setw (bar - visible)
              << ' ';

  std::cout << "\033[0m";

  if (percentage)
    std::cout << " "
              << std::setfill (' ')
              << std::setw (3)
              << (int) (fraction * 100)
              << "%";

  if (elapsed && start != 0)
    std::cout << " "
              << elapsed_time;

  if (estimate && start != 0 && fraction > 0.2)
    std::cout << " "
              << estimate_time;

  std::cout << "\r"
            << std::flush;
}

////////////////////////////////////////////////////////////////////////////////
// Mono style looks like this:
//
// label WWWWWWWWBBBBBBBBBBBBBBBBB  34% 0:12 0:35
//
// ^^^^^                                            Label string
//       ^^^^^^^^                                   Completed bar (in white)
//               ^^^^^^^^^^^^^^^^^                  Incomplete bar (in black)
//                                 ^^^^             Percentage complete
//                                      ^^^^        Elapsed time
//                                           ^^^^   Remaining estimate
void Progress::renderStyleMono ()
{
  // Fraction completed.
  double fraction = (1.0 * (current - minimum)) / (maximum - minimum);

  // Elapsed time.
  time_t now = time (NULL);
  std::string elapsed_time;
  if (elapsed && start != 0)
    elapsed_time = formatTime (now - start);

  // Estimated remaining time.
  std::string estimate_time;
  if (estimate && start != 0)
  {
    if (fraction >= 1e-6)
      estimate_time = formatTime ((time_t) (int) (((now - start) * (1.0 - fraction)) / fraction));
    else
      estimate_time = formatTime (0);
  }

  // Calculate bar width.
  int bar = width
          - (label.length () ? label.length () + 1         : 0)
          - (percentage      ? 5                           : 0)
          - (elapsed         ? elapsed_time.length () + 1  : 0)
          - (estimate        ? estimate_time.length () + 1 : 0);

  if (bar < 1)
    throw std::string ("The specified width is insufficient.");

  int visible = (int) (fraction * bar);

  // Render.
  if (label.length ())
    std::cout << label
              << ' ';

  if (visible > 0)
    std::cout << "\033[47m" // White
              << std::setfill (' ')
              << std::setw (visible)
              << ' ';

  if (bar - visible > 0)
    std::cout << "\033[40m" // Black
              << std::setfill (' ')
              << std::setw (bar - visible)
              << ' ';

  std::cout << "\033[0m";

  if (percentage)
    std::cout << " "
              << std::setfill (' ')
              << std::setw (3)
              << (int) (fraction * 100)
              << "%";

  if (elapsed && start != 0)
    std::cout << " "
              << elapsed_time;

  if (estimate && start != 0)
    std::cout << " "
              << estimate_time;

  std::cout << "\r"
            << std::flush;
}

////////////////////////////////////////////////////////////////////////////////
// Text style looks like this:
//
// label [********                ]  34% 0:12 0:35
//
// ^^^^^                                             Label string
//        ^^^^^^^^                                   Completed bar
//               ^^^^^^^^^^^^^^^^^                   Incomplete bar
//                                  ^^^^             Percentage complete
//                                       ^^^^        Elapsed time
//                                            ^^^^   Remaining estimate
void Progress::renderStyleText ()
{
  // Fraction completed.
  double fraction = (1.0 * (current - minimum)) / (maximum - minimum);

  // Elapsed time.
  time_t now = time (NULL);
  std::string elapsed_time;
  if (elapsed && start != 0)
    elapsed_time = formatTime (now - start);

  // Estimated remaining time.
  std::string estimate_time;
  if (estimate && start != 0)
  {
    if (fraction >= 1e-6)
      estimate_time = formatTime ((time_t) (int) (((now - start) * (1.0 - fraction)) / fraction));
    else
      estimate_time = formatTime (0);
  }

  // Calculate bar width.
  int bar = width
          - 2                                                    // The [ and ]
          - (label.length () ? label.length () + 1         : 0)
          - (percentage      ? 5                           : 0)
          - (elapsed         ? elapsed_time.length () + 1  : 0)
          - (estimate        ? estimate_time.length () + 1 : 0);

  if (bar < 1)
    throw std::string ("The specified width is insufficient.");

  int visible = (int) (fraction * bar);

  // Render.
  if (label.length ())
    std::cout << label
              << ' ';

  std::cout << '[';

  if (visible > 0)
    std::cout << std::setfill ('*')
              << std::setw (visible)
              << '*';

  if (bar - visible > 0)
    std::cout << std::setfill (' ')
              << std::setw (bar - visible)
              << ' ';

  std::cout << ']';

  if (percentage)
    std::cout << " "
              << std::setfill (' ')
              << std::setw (3)
              << (int) (fraction * 100)
              << "%";

  if (elapsed && start != 0)
    std::cout << " "
              << elapsed_time;

  if (estimate && start != 0)
    std::cout << " "
              << estimate_time;

  std::cout << "\r"
            << std::flush;
}

////////////////////////////////////////////////////////////////////////////////


src/vramsteg.cpp
////////////////////////////////////////////////////////////////////////////////
// Vramsteg - Utility for displaying progress bars in shell scripts.
//
// Copyright 2010 - 2015, Paul Beckingham, Federico Hernandez.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
// http://www.opensource.org/licenses/mit-license.php
//
////////////////////////////////////////////////////////////////////////////////

#include <iostream>
#include <sys/ioctl.h>
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <unistd.h>
#include <time.h>
#include <signal.h>
#include <Progress.h>
#include <cmake.h>

extern char *optarg;
extern int optind;
extern int optopt;
extern int opterr;
extern int optreset;

////////////////////////////////////////////////////////////////////////////////
void showUsage ()
{
  std::cout << "\n"
            << "Usage: vramsteg [options]\n"
            << "\n"
            << "  -y, --style <name>          Style of bar rendering\n"
            << "  -l, --label <value>         Progress bar label\n"
            << "  -m, --min <value>           Equivalent to 0%\n"
            << "  -x, --max <value>           Equivalent to 100%\n"
            << "  -c, --current <value>       Current value of progress bar\n"
            << "  -p, --percentage            Show percentage\n"
            << "  -s, --start <value>         Start time epoch\n"
            << "  -w, --width <value>         Width of progress bar, default full width\n"
            << "  -n, --now                   Returns current time as epoch\n"
            << "  -r, --remove                Removes the progress bar\n"
            << "  -e, --elapsed               Show elapsed time (needs --start)\n"
            << "  -t, --estimate              Show estimated remaining time (needs --start)\n"
#ifdef WAITING_FOR_VITAPI
            << "  -d, --done <color>          Color of completed part\n"
            << "  -a, --remaining <color>     Color of incomplete part\n"
#endif
            << "  -v, --version               Show vramsteg version\n"
            << "  -h, --help                  Show command options\n"
            << "\n"
            << "Supported styles are:\n"
            << "  (default)     label \033[42m        \033[41m            \033[0m 40%\n"
            << "  mono          label \033[47m    \033[40m                \033[0m 20%\n"
            << "  text          label [************      ] 60%\n"
            << std::endl;

  exit (0);
}

////////////////////////////////////////////////////////////////////////////////
void showVersion ()
{
  std::cout << "\n"
            << "\033[1m" << PACKAGE_STRING << "\033[0m\n"
            << "Copyright (C) 2010 - 2015, Göteborg Bit Factory\n"
            << "Copyright (C) 2010 - 2015, P. Beckingham, F. Hernandez.\n"
            << "\n"
            << "Vramsteg may be copied only under the terms of the MIT license, "
            << "which may be found in the taskwarrior source kit.\n"
            << "\n"
            << "Documentation for vramsteg can be found using 'man vramsteg', or "
            << "at http://tasktools.org.\n\n"
            << std::flush;

  exit (0);
}

////////////////////////////////////////////////////////////////////////////////
int main (int argc, char** argv)
{
  try
  {
    // We don't want any signals causing the program to quit in mid output, as
    // this would lead to odd colors persisting in the terminal.
    signal (SIGHUP,    SIG_IGN);
    signal (SIGINT,    SIG_IGN);
    signal (SIGKILL,   SIG_IGN);
    signal (SIGPIPE,   SIG_IGN);
    signal (SIGTERM,   SIG_IGN);
    signal (SIGUSR1,   SIG_IGN);
    signal (SIGUSR2,   SIG_IGN);

    long        arg_current    = 0;
#ifdef WAITING_FOR_VITAPI
    std::string arg_done       = "";
#endif
    bool        arg_elapsed    = false;
    bool        arg_estimate   = false;
    std::string arg_label;
    long        arg_max        = 0;
    long        arg_min        = 0;
    bool        arg_percentage = false;
#ifdef WAITING_FOR_VITAPI
    std::string arg_remaining  = "";
#endif
    bool        arg_remove     = false;
    time_t      arg_start      = 0;
    int         arg_width      = 80;
    std::string arg_style      = "";

    // Dynamically determine terminal width.
    unsigned short buff[4];
    if (ioctl (fileno(stdout), TIOCGWINSZ, &buff) != -1)
      arg_width = buff[1];

    static struct option longopts[] = {
      { "current",    required_argument, NULL, 'c' },
#ifdef WAITING_FOR_VITAPI
      { "done",       required_argument, NULL, 'd' },
#endif
      { "elapsed",    no_argument,       NULL, 'e' },
      { "estimate",   no_argument,       NULL, 't' },
      { "label",      required_argument, NULL, 'l' },
      { "max",        required_argument, NULL, 'x' },
      { "min",        required_argument, NULL, 'm' },
      { "now",        no_argument,       NULL, 'n' },
      { "percentage", no_argument,       NULL, 'p' },
#ifdef WAITING_FOR_VITAPI
      { "remaining",  required_argument, NULL, 'a' },
#endif
      { "remove",     no_argument,       NULL, 'r' },
      { "start",      required_argument, NULL, 's' },
      { "version",    no_argument,       NULL, 'v' },
      { "width",      required_argument, NULL, 'w' },
      { "style",      required_argument, NULL, 'y' },
      { "help",       no_argument,       NULL, 'h' },
      { NULL,         0,                 NULL, 0   }
    };

    int ch;
#ifdef WAITING_FOR_VITAPI
    while ((ch = getopt_long (argc, argv, "c:d:etl:x:m:npa:rs:vw:h", longopts, NULL)) != -1)
#else
    while ((ch = getopt_long (argc, argv, "c:etl:x:m:nprs:vw:h", longopts, NULL)) != -1)
#endif
    {
      switch (ch)
      {
      case 'c': arg_current    = atol (optarg);        break;
#ifdef WAITING_FOR_VITAPI
      case 'd': arg_done       = optarg;               break;
#endif
      case 'e': arg_elapsed    = true;                 break;
      case 't': arg_estimate   = true;                 break;
      case 'l': arg_label      = optarg;               break;
      case 'x': arg_max        = atol (optarg);        break;
      case 'm': arg_min        = atol (optarg);        break;
      case 'n': std::cout << time (NULL) << std::endl; exit (0);
      case 'p': arg_percentage = true;                 break;
#ifdef WAITING_FOR_VITAPI
      case 'a': arg_remaining  = optarg;               break;
#endif
      case 'r': arg_remove     = true;                 break;
      case 's': arg_start      = atoi (optarg);        break;
      case 'v': showVersion ();                        break;
      case 'w': arg_width      = atoi (optarg);        break;
      case 'y': arg_style      = optarg;               break;
      case 'h': showUsage ();                          break;

      default:
        std::cout << "<default>" << std::endl;
        break;
      }
    }

    argc -= optind;
    argv += optind;

    // Sanity check arguments.
    if (arg_min || arg_max)
      if (arg_min > arg_max)
        throw std::string ("The --max value must not be less than the --min value.");

    if (arg_min || arg_max || arg_current)
      if (arg_min > arg_current || arg_current > arg_max)
        throw std::string ("The --current value must not lie outside the --min/--max range.");

    if (arg_width && arg_label.length ())
      if (arg_label.length () >= arg_width)
        throw std::string ("The --label string is longer than the allowed --width value.");

    if (! arg_remove && ! (arg_min || arg_current || arg_max))
      showUsage ();

    if (arg_elapsed && arg_start == 0)
      throw std::string ("To use the --elapsed feature, --start must be provided.");

    if (arg_estimate && arg_start == 0)
      throw std::string ("To use the --estimate feature, --start must be provided.");

    // Disallow signals from stopping the program while it is displaying color codes
    // Set up and render Progress object.
    Progress p (arg_label, arg_width, arg_min, arg_max, arg_percentage, arg_remove);
    p.setStyle (arg_style);
    p.setStart (arg_start);
    p.showElapsed (arg_elapsed);
    p.showEstimate (arg_estimate);
    p.removeAfter (arg_remove);
    p.update (arg_current);

    if (arg_remove)
      p.done ();
  }

  catch (const std::string& e) { std::cerr << "Error: " << e << std::endl; }
  catch (...)                  { std::cerr << "Unknown error occurred - please report." << std::endl; }

  return 0;
}

////////////////////////////////////////////////////////////////////////////////

